### PR TITLE
reply: return existing memory ids from extraction

### PR DESCRIPTION
**Pull request by elie222**
# User description
Keep reply-memory learning in one AI extraction step by letting the model point at an existing memory id when a durable idea is already covered.

- change extraction to return either an existing memory id or a nested new-memory payload
- attach evidence directly to existing memories and add focused unit and eval coverage for the new contract

---

# Generated description

Below is a concise technical summary of the changes proposed in this PR:
Extend <code>aiExtractReplyMemoriesFromDraftEdit</code> to return either a <code>matchingExistingMemoryId</code> or a normalized <code>newMemory</code> payload while normalizing writing-style inputs in <code>aiDraftReplyWithConfidence</code>, keeping prompt helpers and eval coverage aligned with the new contract. Ensure <code>syncReplyMemoriesFromDraftSendLogs</code> attaches evidence to matched memories, logs when extraction returns unknown ids, and reports the counts of attached versus created memories.
<table><tr><th>Topic</th><th>Details</th><tr><td><a href=https://baz.co/changes/elie222/inbox-zero/2053?tool=ast&topic=Memory+extraction>Memory extraction</a>
        </td><td>Coordinate memory extraction and drafting prompts by normalizing writing-style inputs, teaching <code>aiExtractReplyMemoriesFromDraftEdit</code> to return either <code>matchingExistingMemoryId</code> or <code>newMemory</code>, updating prompt helpers/pipeline versioning, and covering the new contract through eval and formatting tests focused on created memories.<details><summary>Modified files (5)</summary><ul><li>apps/web/__tests__/eval/reply-memory.test.ts</li>
<li>apps/web/utils/ai/reply/draft-attribution.ts</li>
<li>apps/web/utils/ai/reply/draft-reply.formatting.test.ts</li>
<li>apps/web/utils/ai/reply/draft-reply.ts</li>
<li>apps/web/utils/ai/reply/extract-reply-memories.ts</li></ul></details><details><summary>Latest Contributors(1)</summary><table><tr><th>User</th><th>Commit</th><th>Date</th></tr><tr><td>elie222</td><td>Improve learned writin...</td><td>March 22, 2026</td></tr></table></details></td></tr>
<tr><td><a href=https://baz.co/changes/elie222/inbox-zero/2053?tool=ast&topic=Memory+persistence>Memory persistence</a>
        </td><td>Handle evidence for matched memories by updating <code>syncReplyMemoriesFromDraftSendLogs</code> to attach sources, skip unknown ids with warnings, and report counts, with corresponding reply-memory tests that exercise evidence attachment and unknown-id handling.<details><summary>Modified files (2)</summary><ul><li>apps/web/utils/ai/reply/reply-memory.test.ts</li>
<li>apps/web/utils/ai/reply/reply-memory.ts</li></ul></details><details><summary>Latest Contributors(1)</summary><table><tr><th>User</th><th>Commit</th><th>Date</th></tr><tr><td>elie222</td><td>Improve learned writin...</td><td>March 22, 2026</td></tr></table></details></td></tr></table>
This pull request is reviewed by Baz. Review like a pro on <a href=https://baz.co/changes/elie222/inbox-zero/2053?tool=ast>(Baz)</a>.

### CODE REVIEW

File: apps/web/utils/ai/reply/extract-reply-memories.ts [hunk 1]
```diff
@@ -14,17 +14,33 @@ import type { getEmailAccountWithAi } from "@/utils/user/get";
...
+    const hasNewMemory = Boolean(decision.newMemory);
+
+    if (hasExistingId === hasNewMemory) {
+      ctx.addIssue({
+        code: z.ZodIssueCode.custom,
+        message:
+          "Each reply memory decision must include exactly one of matchingExistingMemoryId or newMemory.",
+      });
+    }
+  });
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/ai/reply/extract-reply-memories.ts]: Fixed. I replaced the `superRefine`-only exclusivity check with a real union shape, so the generated JSON Schema now encodes the one-of contract for `matchingExistingMemoryId` vs `newMemory` instead of relying on a Zod-only runtime refinement.

File: apps/web/utils/ai/reply/extract-reply-memories.ts [hunk 2]
```diff
@@ -91,24 +101,43 @@ export async function aiExtractReplyMemoriesFromDraftEdit({
...
-          ? ""
-          : memory.scopeValue.trim(),
-    }))
-    .filter(
-      (memory) =>
-        memory.scopeType !== ReplyMemoryScopeType.TOPIC ||
-        memory.scopeValue.length > 0,
-    )
+    .map((decision) => {
+      if (decision.matchingExistingMemoryId) {
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/ai/reply/extract-reply-memories.ts:105]: I did not move this validation back into extraction. The sync step already validates `matchingExistingMemoryId` against the fetched `existingMemories` and warns with account/log context before skipping the attach in `reply-memory.ts`, which gives us visibility without silently dropping the model decision earlier. I added a focused regression test covering that path.

File: apps/web/utils/ai/reply/draft-reply.ts [hunk 3]
```diff
@@ -251,7 +251,9 @@ export async function aiDraftReplyWithConfidence({
       : null,
   });
 
-  const effectiveWritingStyle = writingStyle || defaultWritingStyle;
+  const effectiveWritingStyle =
+    writingStyle || learnedWritingStyle || defaultWritingStyle;
+  const advisoryLearnedWritingStyle = writingStyle ? learnedWritingStyle : null;
 
   const prompt = getUserPrompt({
     messages,
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/ai/reply/draft-reply.ts:267]: Fixed in `6ce46ea`. `writingStyle` and `learnedWritingStyle` are now trimmed before precedence selection, whitespace-only values are treated as empty, and the prompt omits empty style blocks. I also added focused formatting tests for whitespace-only explicit and learned style values.

**Comment by elie222** [comment]
Post-merge follow-up tracked in Linear: INB-179\n\nReview reply memory rebuild after the March 27, 2026 reset and prompt updates on March 30, 2026:\nhttps://linear.app/inbox-zero-ai/issue/INB-179/review-reply-memory-rebuild-after-reset
